xla/service/gpu/gpu_compiler.cc

```diff
@@ -884,6 +868,44 @@ absl::Status RunCollectiveOptimizationPasses(
         /*reuse_pipelined_op_buffer=*/HloPredicateFalse};
     collectives_pipeline.AddPass<CollectivePipeliner>(config);
   }
+  return absl::OkStatus();
+}
+
+absl::Status RunPostLayoutCollectivePipelinerPasses(
+    HloModule* hlo_module) {
+  const DebugOptions& debug_options = hlo_module->config().debug_options();
+  HloPassPipeline collectives_pipeline("collective-pipeliner-optimizations");
+  if (debug_options.xla_gpu_run_post_layout_collective_pipeliner()) {
+    TF_RETURN_IF_ERROR(
+        AddCollectivePipelinerPasses(debug_options, collectives_pipeline));
+    collectives_pipeline.AddPass<WhileLoopTripCountAnnotator>();
```

Comment:
Why does `WhileLoopTripCountAnnotator` appear here and not in the other case?

Reply:
We also call WhileLoopTripCountAnnotator at the end of the collective pipeline, i think any pass that changes the loop trip count would have to call this after it.